File: cs/Minsk/CodeAnalysis/Binding/BoundBinaryOperator.cs
using Minsk.CodeAnalysis.Syntax;

namespace Minsk.CodeAnalysis.Binding;

internal sealed class BoundBinaryOperator
{
    private static readonly BoundBinaryOperator[] KnownOperators =
    {
        new(SyntaxKind.PlusToken, BoundBinaryOperatorKind.Addition, typeof(int), typeof(int), typeof(int)),
        new(SyntaxKind.MinusToken, BoundBinaryOperatorKind.Subtraction, typeof(int), typeof(int), typeof(int)),
        new(SyntaxKind.StarToken, BoundBinaryOperatorKind.Multiplication, typeof(int), typeof(int), typeof(int)),
        new(SyntaxKind.SlashToken, BoundBinaryOperatorKind.Division, typeof(int), typeof(int), typeof(int)),
        new(
            SyntaxKind.AmpersandAmpersandToken,
            BoundBinaryOperatorKind.LogicalAnd,
            typeof(bool),
            typeof(bool),
            typeof(bool)
        ),
        new(
            SyntaxKind.PipePipeToken,
            BoundBinaryOperatorKind.LogicalOr,
            typeof(bool),
            typeof(bool),
            typeof(bool)
        ),
        new(
            SyntaxKind.EqualsEqualsToken,
            BoundBinaryOperatorKind.Equality,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
        new(
            SyntaxKind.BangEqualsToken,
            BoundBinaryOperatorKind.Inequality,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
        new(
            SyntaxKind.EqualsEqualsToken,
            BoundBinaryOperatorKind.Equality,
            typeof(bool),
            typeof(bool),
            typeof(bool)
        ),
        new(
            SyntaxKind.BangEqualsToken,
            BoundBinaryOperatorKind.Inequality,
            typeof(bool),
            typeof(bool),
            typeof(bool)
        ),
        new(
            SyntaxKind.LessToken,
            BoundBinaryOperatorKind.LessThan,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
        new(
            SyntaxKind.LessEqualsToken,
            BoundBinaryOperatorKind.LessThanOrEqual,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
        new(
            SyntaxKind.GreaterToken,
            BoundBinaryOperatorKind.GreaterThan,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
        new(
            SyntaxKind.GreaterEqualsToken,
            BoundBinaryOperatorKind.GreaterThanOrEqual,
            typeof(int),
            typeof(int),
            typeof(bool)
        ),
    };

    private BoundBinaryOperator(
        SyntaxKind syntaxKind,
        BoundBinaryOperatorKind operatorKind,
        Type leftType,
        Type rightType,
        Type resultType
    )
    {
        SyntaxKind = syntaxKind;
        OperatorKind = operatorKind;
        LeftType = leftType;
        RightType = rightType;
        ResultType = resultType;
    }

    private Type LeftType { get; }
    private SyntaxKind SyntaxKind { get; }
    public BoundBinaryOperatorKind OperatorKind { get; }
    private Type RightType { get; }
    public Type ResultType { get; }

    public static BoundBinaryOperator? BindBinaryOperator(
        BoundExpression left,
        SyntaxKind operatorKind,
        BoundExpression right
    )
    {
        return KnownOperators.FirstOrDefault(knownOperator =>
            knownOperator.LeftType == left.Type && knownOperator.SyntaxKind == operatorKind &&
            knownOperator.RightType == right.Type);
    }
}

File: nim/src/minskpkg/codeAnalysis/binding/expressions/boundBinaryOperator.nim
import std/options

import minskpkg/minskObject
import minskpkg/codeAnalysis/syntax/syntaxKind
import sequtilsExt/sequtilsExt

type
  BoundBinaryOperatorKind* {.pure.} = enum
    Addition
    Subtraction
    Multiplication
    Division
    LogicalAnd
    LogicalOr
    Equality
    Inequality

  BoundBinaryOperator* = object
    syntaxKind: SyntaxKind
    kind*: BoundBinaryOperatorKind
    leftType: MinskObjectKind
    rightType: MinskObjectKind
    resultType*: MinskObjectKind

const
  BOUND_BINARY_OPERATORS = [
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.PlusToken,
      kind: BoundBinaryOperatorKind.Addition,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokInteger,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.MinusToken,
      kind: BoundBinaryOperatorKind.Subtraction,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokInteger,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.StarToken,
      kind: BoundBinaryOperatorKind.Multiplication,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokInteger,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.SlashToken,
      kind: BoundBinaryOperatorKind.Division,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokInteger,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.AmpersandAmpersandToken,
      kind: BoundBinaryOperatorKind.LogicalAnd,
      leftType: mokBoolean,
      rightType: mokBoolean,
      resultType: mokBoolean,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.PipePipeToken,
      kind: BoundBinaryOperatorKind.LogicalOr,
      leftType: mokBoolean,
      rightType: mokBoolean,
      resultType: mokBoolean,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.EqualsEqualsToken,
      kind: BoundBinaryOperatorKind.Equality,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokBoolean,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.BangEqualsToken,
      kind: BoundBinaryOperatorKind.Inequality,
      leftType: mokInteger,
      rightType: mokInteger,
      resultType: mokBoolean,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.EqualsEqualsToken,
      kind: BoundBinaryOperatorKind.Equality,
      leftType: mokBoolean,
      rightType: mokBoolean,
      resultType: mokBoolean,
    ),
    BoundBinaryOperator(
      syntaxKind: SyntaxKind.BangEqualsToken,
      kind: BoundBinaryOperatorKind.Inequality,
      leftType: mokBoolean,
      rightType: mokBoolean,
      resultType: mokBoolean,
    ),
  ]

proc bindBinaryOperator*(
  kind: SyntaxKind,
  leftType: MinskObjectKind,
  rightType: MinskObjectKind
): Option[BoundBinaryOperator] =
  sequtilsExt.find(
    BOUND_BINARY_OPERATORS,
    proc (op: BoundBinaryOperator): bool =
      op.syntaxKind == kind and op.leftType == leftType and op.rightType == rightType
  )
